### PR TITLE
Hash should not default to 11 if it is specified

### PR DESCRIPTION
**Pull request by dylandrop**
`Hash#new` has some strange behavior of setting the initial capacity, ignoring what is specified, if the specified value is less than 11. Instead, it should let the user set the capacity to whatever they want, but complain if a non-positive value is specified.

### CODE REVIEW

**Comment by spalladino** [comment]
My guess is that it was set up this way since the difference in size between 1-10 is negligible, and creating a hash as small as under 10 elements can yield terrible performance. @asterite do you want to chime in?

**Comment by asterite** [comment]
The capacity must always match values in `HASH_PRIMES` to get less hash collitions, if I remember correctly

**Comment by asterite** [comment]
However, it's OK to raise an ArgumentError if the capacity is negative. And I guess we should cap the initialize capacity to one of the prime values (right now that's not the case), though I'm not sure that's really needed.

**Comment by spalladino** [comment]
Cap the initial capacity to one of the primes, or actually try to match it to the closest largest one? For instance, if a user creates a hash with initial capacity 12, should we actually create a hash with size 19?

**Comment by asterite** [comment]
I don't know... A hash literal will set the initial capacity to the number of elements it has. If the initial capacity is less than what is passed there will be a rehash of all the values, which is not good. So I think using a larger value is better.

**Comment by spalladino** [comment]
Ok, then we should:
- Create an issue to use the next largest prime instead of the actual inicial capacity set by the user
- Kindly ask @dylandrop to keep the original `if < 11 then 11`, and submit this PR with the exception on negative capacity only

Agree?

**Comment by dylandrop** [comment]
> Kindly ask @dylandrop to keep the original if < 11 then 11, and submit this PR with the exception on negative capacity only

Happy to do this, but want to add: I think we should at least warn the user about this rule with a message. Or, I can add documentation for it.

**Comment by spalladino** [comment]
I'd avoid a warning on runtime for that, but I agree on adding it to the documentation; however, I'd wait to define whether we want to just keep the current behaviour, or go for always using the next largest prime.

**Comment by spalladino** [comment]
We went through the Ruby implementation, and it seems that when a capacity is passed in, Ruby looks for the largest power of 2 smaller than the requested value, and uses the prime generated from that power of 2 as the actual capacity. See [here](https://github.com/ruby/ruby/blob/v1_9_3_551/st.c#L132-L134) for the code.

I'd suggest we copy that implementation, and add a notification about it on the docs as @dylandrop suggested; as well as raise the exception if the capacity is negative.

@dylandrop this is actually exceeding the original scope of this PR. Would you like to add this implementation here, or would you rather close this PR and we open a new issue for you or someone else to work on it?

**Comment by dylandrop** [comment]
@spalladino No I can take care of it -- seems reasonable.

**Comment by spalladino** [comment]
Awesome, thanks a lot!!

**Comment by dylandrop** [comment]
@spalladino @asterite Alright, we're good for another look.

File: spec/std/hash_spec.cr [hunk 1]
```diff
@@ -805,20 +811,25 @@ describe "Hash" do
     items.uniq.size
   end
 
-  it "creates with initial capacity" do
+  it "sets the initial capacity when not passed" do
+    hash = Hash(Int32, Int32).new
+    hash.@buckets_size.should eq(PRIMES[0])
```
Comment (Sija) [inline spec/std/hash_spec.cr]: `PRIMES` -> `Hash::PRIMES`

File: src/hash.cr [hunk 2]
```diff
@@ -834,6 +864,50 @@ class Hash(K, V)
     find_entry_in_bucket entry, key
   end
 
+  # Table of prime numbers 2^n+a, 2<=n<=30.
+  PRIMES = [
```
Comment (Sija) [inline src/hash.cr]: This can safely be a `Tuple`.

Comment (ysbaddaden) [inline src/hash.cr]: Or a StaticArray; all elements are the same type (Int32).

```crystal
Int32.static_array(8 + 3, 16 + 3, ...)
```

**Comment by dylandrop** [comment]
Interesting, it seems from the [Travis build](https://travis-ci.org/crystal-lang/crystal/jobs/207764883) that using this approach leads to some very large memory allocation, causing the process to crash. I'm wondering if I should just use the size the caller passed in if the passed size exceeds a certain value (say `8192 + 27`).

Or, perhaps there's some bug in my code I'm not seeing.

Output at the bottom of the Travis job:

```
GC Warning: Repeated allocation of very large block (appr. size 268439552):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 268439552):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 1073745920):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 2147487744):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 33558528):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 67112960):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 134221824):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 268439552):
	May lead to memory leak and poor performance
```

File: src/hash.cr [hunk 3]
```diff
@@ -13,8 +13,8 @@ class Hash(K, V)
 
   def initialize(block : (Hash(K, V), K -> V)? = nil, initial_capacity = nil)
     initial_capacity ||= 11
-    initial_capacity = 11 if initial_capacity < 11
     initial_capacity = initial_capacity.to_i
+    raise ArgumentError.new("Hash capacity must be positive") if initial_capacity <= 0
```
Comment (bew) [inline src/hash.cr]: I don't think it should raise if `initial_capacity == 0`

The user might want to create a Hash with initial capacity of 0. Even if it seems pointless, there might be some cases where it is usefull!

Comment (ozra) [inline src/hash.cr]: Agree

**Comment by akzhan** [comment]
For now I thinking about another implementation of `Hash` itself, using open addressing and powers of two.

Ruby 2.4 increases its Hash performance for 10-200% with this change (see https://bugs.ruby-lang.org/issues/12142#note-142 for details).

**Comment by RX14** [comment]
@akzhan I'm sure a PR using open addressing would be accepted, it's been discussed a few times for sure.

**Comment by akzhan** [comment]
Just asked @funny-falcon for some details (its implementation looks cleaner for me).

**Comment by funny-falcon** [comment]
I'd be happy to make PR with open-addressing implementation.
Though I'm not very organized man. So if i will not create PR within 3 weeks, better do it yourself.

**Comment by akzhan** [comment]
@funny-falcon Feel free to publish clean idea implemented in C++, it's easy to convert good C++ code to Crystal.

**Comment by RX14** [comment]
See this blog post when implementing a hash table: http://www.idryman.org/blog/2017/05/03/writing-a-damn-fast-hash-table-with-tiny-memory-footprints/

**Comment by RX14** [comment]
Also make sure to read the [HN comments](https://news.ycombinator.com/item?id=14290055) and the references at the end of the article. There's a load of good resources.

**Comment by funny-falcon** [comment]
@RX14 , I will try to reimplement same approach, used for Ruby. Perhaps it is not the best.

I think, there could be concurrent PR with different implementations, so you can implement other one.

@akzhan , there is interest to use C++ :-) I don't know Crystal well yet, so it will be good reason to learn more.

**Comment by akzhan** [comment]
@funny-falcon Please mention #4557 as base issue for pull request, thanks.

**Comment by straight-shoota** [comment]
This has been fixed in #8017

**Pull request closed without merging**